packages/material-ui-lab/src/Timeline/Timeline.tsx

```diff
@@ -0,0 +1,106 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import clsx from 'clsx';
+import { InternalStandardProps as StandardProps } from '@material-ui/core';
+import { capitalize } from '@material-ui/core/utils';
+import { withStyles, createStyles, WithStyles } from '@material-ui/core/styles';
+import TimelineContext from './TimelineContext';
+
+export const styles = createStyles({
+  /* Styles applied to the root element. */
+  root: {
+    display: 'flex',
+    flexDirection: 'column',
+    padding: '6px 16px',
+    flexGrow: 1,
+  },
+  /* Styles applied to the root element if `align="left"`. */
+  alignLeft: {},
+  /* Styles applied to the root element if `align="right"`. */
+  alignRight: {},
+  /* Styles applied to the root element if `align="alternate"`. */
+  alignAlternate: {},
+});
+
+export type TimelineClassKey = keyof WithStyles<typeof styles>['classes'];
+
+export interface TimelineProps extends StandardProps<React.HTMLAttributes<HTMLUListElement>> {
+  /**
+   * The position where the timeline's content should appear.
+   * @default 'left'
+   */
+  align?: 'left' | 'right' | 'alternate';
+  /**
+   * The content of the component.
+   */
+  children?: React.ReactNode;
+  /**
+   * Override or extend the styles applied to the component.
+   */
+  classes?: {
+    /** Styles applied to the root element. */
+    root?: string;
+    /** Styles applied to the root element if `align="left"`. */
+    alignLeft?: string;
+    /** Styles applied to the root element if `align="right"`. */
+    alignRight?: string;
+    /** Styles applied to the root element if `align="alternate"`. */
+    alignAlternate?: string;
+  };
+}
+
+const Timeline = React.forwardRef<HTMLUListElement, TimelineProps>(function Timeline(props, ref) {
+  const { align = 'left', classes, className, ...other } = props;
+
+  return (
+    <TimelineContext.Provider value={{ align }}>
+      <ul
+        className={clsx(
+          classes!.root,
```

Comment:
I'm not super happy about this one, but it seems to be a reasonable tradeoff.

Reply:
You're only typing the outer props i.e. the props a user can pass in but these are not the same props the component actually receives due to higher-order components.

We need to come up with a robust approach once we're starting to switch to `emotion` so that we don't degrade user types. Not important for this PR but v5 release.

Reply:
@eps1lon Agree, the main motivation for opening this pull request was to help with #22692.